src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,323 @@
+    }
+
+    /**
+     * Determines if the keywordIndex-th keyword is confirmed i.e. user has typed a space after it
+     * Assumption: userInput has at at least keywordIndex+1 keywords, separated by whitespace.
+     * @param userInput
+     * @param keywordIndex
+     * @return
+     */
+    private static boolean isConfirmedKeyword(String userInput, int keywordIndex) {
+        return !(keywordIndex == userInput.split("\\s+").length - 1 && !userInput.endsWith(" "));
+    }
+
+    private static String getMatchedLabelName(List<String> repoLabels, String keyword) {
+        List<String> newMatchedLabels = new ArrayList<>();
+        newMatchedLabels.addAll(repoLabels);
+        newMatchedLabels = filterByName(newMatchedLabels, getName(keyword));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(keyword));
+        return newMatchedLabels.get(0);
+    }
+
+    private static boolean hasExactlyOneMatchedLabel(List<String> repoLabels, String keyword) {
+        List<String> newMatchedLabels = new ArrayList<>();
+        newMatchedLabels.addAll(repoLabels);
+        newMatchedLabels = filterByName(newMatchedLabels, getName(keyword));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(keyword));
+        return newMatchedLabels.size() == 1;
+    }
+
+    private static boolean hasExclusiveGroup(String name) {
+        return TurboLabel.getDelimiter(name).isPresent() && TurboLabel.getDelimiter(name).get().equals(".");
+    }
+
+    private static String getGroup(String labelName) {
+        if (!hasGroup(labelName)) return "";
+
+        return labelName.substring(0, labelName.indexOf(TurboLabel.getDelimiter(labelName).get()));
+    }
+
+    private static String getName(String labelName) {
+        if (!hasGroup(labelName)) return labelName;
+
+        return labelName.substring(labelName.indexOf(TurboLabel.getDelimiter(labelName).get()) + 1);
+    }
+
+    private static boolean hasGroup(String name) {
+        return TurboLabel.getDelimiter(name).isPresent();
+    }
+
+    private static List<String> convertToList(Set<String> labelSet){
+        return new ArrayList<>(labelSet);
+    }
+
+    private static List<String> filterByName(List<String> repoLabels, String labelName) {
+        return repoLabels
+                .stream()
+                .filter(name -> Utility.containsIgnoreCase(getName(name), labelName))
+                .collect(Collectors.toList());
+    }
+
+    private static List<String> filterByGroup(List<String> repoLabels, String labelGroup) {
+        if (labelGroup.isEmpty()) return repoLabels;
+
+        return repoLabels
+                .stream()
+                .filter(name -> {
+                    if (hasGroup(name)) {
+                        return Utility.containsIgnoreCase(getGroup(name), labelGroup);
+                    }
+                    return false;
+                })
+                .collect(Collectors.toList());
+    }
+}
```

Comment:
1. Add header comments to privates methods. The only time to omit the header comment is when the header comment has nothing more than what is given in the method signature.
2. Push label-specific utility functions to `TurboLabel` class.